source/time_stepping/interface.cc

```diff
@@ -108,13 +113,96 @@ namespace aspect
         new_time_step = std::min(new_time_step, this->get_parameters().maximum_first_time_step);
 
       // Make sure we reduce the time step length appropriately if we terminate after this step
-      new_time_step = termination_manager.check_for_last_time_step(new_time_step);
+      TimeStepInfo info;
+      info.reduced_by_termination_plugin = false;
+      {
+        const double reduced_time_step = termination_manager.check_for_last_time_step(new_time_step);
+        if (reduced_time_step < new_time_step)
+          {
+            new_time_step = reduced_time_step;
+            info.reduced_by_termination_plugin = true;
+          }
+      }
+      info.next_time_step_size = new_time_step;
 
       AssertThrow (new_time_step > 0,
                    ExcMessage("The time step length for the each time step needs to be positive, "
                               "but the computed step length was: " + std::to_string(new_time_step) + ". "
                               "Please check the time stepping plugins in use."));
-      return new_time_step;
+
+      Reaction reaction = Reaction::advance;
+      double repeat_step_size = std::numeric_limits<double>::max();
+
+      for (const auto &plugin : active_plugins)
+        {
+          const std::pair<Reaction, double> answer = plugin->determine_reaction(info);
+          reaction = static_cast<Reaction>(std::min(reaction, answer.first));
+          repeat_step_size = std::min(repeat_step_size, answer.second);
+        }
+
+      reaction = static_cast<Reaction>(Utilities::MPI::min(static_cast<int>(reaction), this->get_mpi_communicator()));
+      repeat_step_size = Utilities::MPI::min(repeat_step_size, this->get_mpi_communicator());
+
+      this->current_reaction = reaction;
```

Comment:
Hmm, not sure why you use `this->` here and in the lines below. To make clear that they are class members?

Reply:
I find it more readable while working on the code, but I think it is not our usual style, so I will delete it.